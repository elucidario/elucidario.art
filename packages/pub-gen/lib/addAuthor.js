import path from "path";
import { getPaths, createFile } from "@elucidario/pkg-paths";
import { Console } from "@elucidario/pkg-console";

import { pubGenPrompt } from "./prompt/pubGenPrompt.js";
import {
    makeInquirer,
    pubGenConfig,
    packageJson,
    cleanFalsy,
} from "./utils.js";

let authors = [];

export const addAuthor = async (args) => {
    const { publication } = args;
    const pkg = packageJson(publication);
    const console = new Console(pkg);

    await makeInquirer({
        title: "Adicionando autor...",
        publication,
        prompt: pubGenPrompt("addAuthor", args),
        callback: async (answers) => {
            const { contributor, addMoreAuthor } = cleanFalsy(answers);

            authors.push(contributor);

            if (addMoreAuthor) {
                console.success({
                    message: `Autor ${contributor.title} adicionado com sucesso! Adicione mais autores:`,
                });
                addAuthor(args);
            } else {
                const pubGenJson = pubGenConfig(publication);

                pubGenJson.contributors = [
                    ...pubGenJson.contributors,
                    ...authors,
                ];

                createFile(
                    {
                        filePath: path.resolve(
                            getPaths().publications,
                            publication,
                            "pub-gen.json"
                        ),
                        ext: "json",
                    },
                    pubGenJson
                );

                const authorsNames = authors
                    .map((author) => author.title)
                    .join(", ");

                console.success({
                    message: `Autores adicionados com sucesso: ${authorsNames}`,
                });

                authors = [];
            }
        },
    });
};
