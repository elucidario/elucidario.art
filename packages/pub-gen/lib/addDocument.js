"use strict";

import fs from "fs";
import path, { dirname } from "path";
import inquirer from "inquirer";
import chalk from "chalk";

import lodash from "lodash";

import { createDoc } from "@elucidario/pkg-md-to-gdoc";

import { pubGenPrompt } from "./prompt.js";
import { getCredentials } from "./getCredentials.js";

const { merge } = lodash;

const __dirname = path.resolve();

export const addDocument = async (args) => {
    const rootPath = path.resolve(__dirname, "..", "..");
    const templatePath = path.resolve(
        rootPath,
        "packages",
        "pub-gen",
        "template"
    );
    const publication = args.publication;
    const pubPath = path.resolve(rootPath, "publications");
    const dirName = path.resolve(pubPath, publication);

    inquirer.prompt(pubGenPrompt("addDocument", args)).then(async (answers) => {
        const { anotherDocument, document } = answers;

        let documents = [];
        if (anotherDocument) {
            console.log(
                chalk.green(`${document.title} adicionado com sucesso`)
            );
            documents = await createGoogleIntegration(
                document,
                templatePath,
                dirName,
                rootPath
            );
            addDocument(args);
        } else {
            documents = await createGoogleIntegration(
                document,
                templatePath,
                dirName,
                rootPath
            );
            /**
             * Save pub-gen.json
             */
            saveDocumentsToPubGenJson(documents, dirName);
        }
    });
};

const saveDocumentsToPubGenJson = (documents, dirName) => {
    const pubGenJson = JSON.parse(
        fs.readFileSync(path.resolve(dirName, "pub-gen.json"))
    );
    pubGenJson.documents = [...pubGenJson.documents, ...documents];
    try {
        fs.writeFileSync(
            path.resolve(dirName, "pub-gen.json"),
            JSON.stringify(pubGenJson, null, 4)
        );
        const documentsNames = documents.map((doc) => doc.title).join(", ");
        console.log(
            chalk.green(`Successfully create documents: ${documentsNames}`)
        );
    } catch (error) {
        console.log(chalk.red("It was not possible to save pub-gen.json"));
        console.error(error);
    }
};

const documents = [];
const createGoogleIntegration = async (document, from, to, rootPath) => {
    try {
        /**
         * Try to get credentials
         */
        let credentials = {};
        try {
            credentials = getCredentials();
        } catch (error) {
            console.log(chalk.red("It was not possible to get credentials"));
            console.error(error);
            return error;
        }

        switch (document.type) {
            case "doc":
                /**
                 * Create references directory and copy demo references
                 */
                createReferences(document.preset, from, to);

                /**
                 * Create content directory and copy demo content
                 */
                createDemoContent(document.language, from, to);

                /**
                 * Create Google Docs
                 */
                try {
                    const defaultDoc = {
                        ...document,
                        keywords: document.keywords
                            .split(",")
                            .map((keyword) => keyword.trim()),
                    };
                    const response = await createDoc(
                        document.title,
                        credentials,
                        rootPath
                    );
                    const doc = merge({}, defaultDoc, response);
                    documents.push(doc);
                    return documents;
                } catch (error) {
                    console.log(
                        chalk.red("It was not possible to create Google Docs")
                    );
                    console.error(error);
                    return error;
                }

            case "slide":
                console.warn(chalk.yellowBright("Slides not implemented yet"));
                // docs = await createDoc(
                //     document.title,
                //     credentials,
                //     rootPath
                // );
                // console.log(docs);
                throw new Error("Slides not implemented yet");

            case "sheet":
                console.warn(chalk.yellowBright("Sheets not implemented yet"));
                // docs = await createDoc(
                //     document.title,
                //     credentials,
                //     rootPath
                // );
                // console.log(docs);
                throw new Error("Sheets not implemented yet");

            default:
                throw new Error("Document type not found");
        }
    } catch (error) {
        console.log(
            chalk.red("It was not possible to create integration with Google")
        );
        console.error(error);
        return error;
    }
};

const createReferences = (preset, from, to) => {
    try {
        let demoRefs = fs.readdirSync(path.resolve(from, "references", preset));
        if (demoRefs.length > 0) {
            fs.mkdirSync(path.resolve(to, "references", preset), {
                recursive: true,
            });
            demoRefs.forEach((file) => {
                fs.copyFileSync(
                    path.resolve(from, "references", preset, file),
                    path.resolve(to, "references", preset, file)
                );
            });
        }
        console.log(chalk.green("Successfully created references."));
    } catch (error) {
        console.log(chalk.red("It was not possible to create references"));
        console.error(error);
    }
};

const createDemoContent = (language, from, to) => {
    try {
        const demoContent = fs.readdirSync(path.resolve(from, "content"));
        fs.mkdirSync(path.resolve(to, "content", language), {
            recursive: true,
        });
        demoContent.forEach((file) => {
            fs.copyFileSync(
                path.resolve(from, "content", file),
                path.resolve(to, "content", language, file)
            );
        });
        console.log(chalk.green("Successfully created demo content."));
    } catch (error) {
        console.log(chalk.red("It was not possible to create demo content"));
        console.error(error);
    }
};
