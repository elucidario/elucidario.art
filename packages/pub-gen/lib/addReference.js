import {
    authorPrompt,
    datePrompt,
    referencePrompt,
    authorProperties,
    dateProperties,
} from "./prompt/referencePrompt.js";
import inquirer from "inquirer";

import { Console } from "@elucidario/pkg-console";
import packageJson from "../package.json" assert { type: "json" };
import { kebabCase } from "lodash-es";

const console = new Console(packageJson);

export const addReference = async (defaults) => {
    try {
        console.log("Selecione o tipo de referencia:");
        inquirer
            .prompt(referencePrompt("create", defaults))
            .then(async (answers) => {
                await addReferenceByType(answers.type, defaults);
            });
    } catch (error) {
        console.error(error);
    }
};

export const addReferenceByType = async (type, defaults) => {
    try {
        console.log(`Adicionando nova referencia do tipo ${type}`);
        await inquirer
            .prompt(referencePrompt(type, defaults))
            .then(async (answers) => {
                // parse answers and return a new object with only the non-empty values
                let reference = Object.keys(answers[type]).reduce(
                    (acc, cur) => {
                        if (answers[type][cur].length) {
                            acc[cur] = answers[type][cur];
                        }
                        return acc;
                    },
                    {}
                );
                reference["type"] = type;
                reference["_slug"] = kebabCase(reference.title);

                // configura a referencia
                try {
                    // configura os autores
                    try {
                        console.log("Selecione o tipo de autor:");
                        await inquirer
                            .prompt(authorPrompt("type", { defaults }))
                            .then(async (answers) => {
                                const { type } = answers;

                                // configura os nomes dos autores
                                try {
                                    const authors = await author(
                                        type,
                                        defaults
                                    ).then((authorNames) => {
                                        // parse authorNames and return a new object with only the non-empty values
                                        return Object.keys(authorNames).reduce(
                                            (acc, cur) => {
                                                if (authorNames[cur].length) {
                                                    acc[cur] = authorNames[cur];
                                                }
                                                return acc;
                                            },
                                            {}
                                        );
                                    });

                                    reference = { ...reference, ...authors };
                                } catch (error) {
                                    console.error(error);
                                    throw new Error(error);
                                }
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as datas
                    try {
                        console.log("Selecione o tipo de data:");
                        await inquirer
                            .prompt(datePrompt("type", { defaults }))
                            .then(async (answers) => {
                                const { type } = answers;

                                // configura os tipos de datas
                                try {
                                    const dates = await date(
                                        type,
                                        defaults
                                    ).then((dateNames) => {
                                        // parse authorNames and return a new object with only the non-empty values
                                        return Object.keys(dateNames).reduce(
                                            (acc, cur) => {
                                                if (
                                                    Object.keys(dateNames[cur])
                                                        .length
                                                ) {
                                                    acc[cur] = dateNames[cur];
                                                }
                                                return acc;
                                            },
                                            {}
                                        );
                                    });

                                    reference = { ...reference, ...dates };
                                } catch (error) {
                                    console.error(error);
                                    throw new Error(error);
                                }
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as propriedades comuns
                    try {
                        console.log("Selecione as propriedades comuns:");
                        await inquirer
                            .prompt(
                                referencePrompt("common", {
                                    defaults,
                                })
                            )
                            .then(async (answers) => {
                                const cleanAnswers = Object.keys(
                                    answers
                                ).reduce((acc, cur) => {
                                    if (answers[cur].length) {
                                        acc[cur] = answers[cur];
                                    }
                                    return acc;
                                }, {});

                                reference = {
                                    ...reference,
                                    ...cleanAnswers,
                                };
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as propriedades de arquivo
                    try {
                        console.log("Selecione as propriedades de arquivo:");
                        await inquirer
                            .prompt(
                                referencePrompt("archive", {
                                    defaults,
                                })
                            )
                            .then(async (answers) => {
                                const cleanAnswers = Object.keys(
                                    answers
                                ).reduce((acc, cur) => {
                                    if (answers[cur].length) {
                                        acc[cur] = answers[cur];
                                    }
                                    return acc;
                                }, {});

                                reference = {
                                    ...reference,
                                    ...cleanAnswers,
                                };
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }
                } catch (error) {
                    console.error(error);
                    throw new Error(error);
                }

                console.log({ reference }, true);
            });
    } catch (error) {
        throw new Error(error);
    }
};

// cria um objeto em que cada propriedade é um item do array authorProperties
const authorNames = authorProperties.reduce((acc, cur) => {
    acc[cur] = [];
    return acc;
}, {});

export const author = async (type, { defaults }) => {
    console.log(`Adicionando novo autor do tipo ${type}`);
    await inquirer
        .prompt(
            authorPrompt("names", {
                defaults,
                name: type,
            })
        )
        .then(async (answers) => {
            const { addMore, ...definedAuthor } = answers;
            const cleanAuthor = Object.keys(definedAuthor[type]).reduce(
                (acc, cur) => {
                    if (definedAuthor[type][cur]) {
                        acc[cur] = definedAuthor[type][cur];
                    }
                    return acc;
                },
                {}
            );
            authorNames[type].push(cleanAuthor);
            if (addMore) {
                await inquirer
                    .prompt(
                        authorPrompt("type", {
                            defaults,
                        })
                    )
                    .then(async (answers) => {
                        const { type } = answers;
                        await author(type, { defaults });
                    });
            }
        });

    return authorNames;
};

// cria um objeto em que cada propriedade é um item do array dateProperties
const dateNames = dateProperties.reduce((acc, cur) => {
    acc[cur] = {};
    return acc;
}, {});
export const date = async (type, { defaults }) => {
    console.log(`Adicionando nova data do tipo ${type}`);
    await inquirer
        .prompt(
            datePrompt("date", {
                defaults,
                name: type,
            })
        )
        .then(async (answers) => {
            const { addMore, ...definedDate } = answers;

            const cleanDate = Object.keys(definedDate[type]).reduce(
                (acc, cur) => {
                    if (definedDate[type][cur]) {
                        acc[cur] = definedDate[type][cur];
                    }
                    return acc;
                },
                {}
            );
            dateNames[type] = cleanDate;
            if (addMore) {
                await inquirer
                    .prompt(
                        datePrompt("type", {
                            defaults,
                            name: type,
                        })
                    )
                    .then(async (answers) => {
                        const { type } = answers;
                        await date(type, { defaults, name: type });
                    });
            }
        });

    return dateNames;
};
