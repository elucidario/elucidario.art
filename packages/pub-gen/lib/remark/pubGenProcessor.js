import { unified } from "unified";
import markdown from "remark-parse";
import remarkGfm from "remark-gfm";
import unifiedPrettier from "unified-prettier";
import remarkFrontmatter from "remark-frontmatter";
import remarkStringify from "remark-stringify";
import { citePlugin } from "@benrbray/remark-cite";
import { Console } from "@elucidario/pkg-console";

import tableParser from "./table/tableParser.js";
import codeParser from "./code/codeParser.js";
import embedParser from "./embed/embedParser.js";
import citeParser from "./cite/citeParser.js";
import countParser from "./count/countParser.js";
import mermaidParser from "./mermaid/mermaidParser.js";

export const pubGenProcessor = async (content, options) => {
    if (!content) return Promise.reject(new Error("No content provided."));
    if (!options) return Promise.reject(new Error("No options provided."));
    if (typeof options.publication === "undefined")
        return Promise.reject(new Error("No publication provided."));
    if (typeof options.lang === "undefined")
        return Promise.reject(new Error("No lang provided."));
    if (typeof options.style === "undefined")
        return Promise.reject(new Error("No style provided."));
    if (typeof options.assets === "undefined")
        return Promise.reject(new Error("No assets provided."));
    if (typeof options.assetsTitles === "undefined")
        return Promise.reject(new Error("No assetsTitles provided."));
    if (typeof options.pkg === "undefined")
        return Promise.reject(new Error("No package.json provided."));

    const console = new Console(options.pkg);

    try {
        const plugins = {
            markdown: {
                commonmark: true,
                footnotes: true,
                pedantic: true,
            },
            frontmatter: {
                type: "yaml",
                fence: "---",
            },
            cite: {},
            pubGen: {
                ...options,
            },
            stringfy: {},
            prettier: {},
        };

        const usedPlugins = [
            [markdown, plugins.markdown], // always on.
            [remarkGfm, plugins.markdown], // always on.
            [remarkFrontmatter, plugins.frontmatter], // always on.

            [tableParser, plugins.pubGen], // pub-gen table parser.
            [codeParser, plugins.pubGen], // pub-gen code parser.
            [embedParser, plugins.pubGen], // pub-gen embed parser.
            [mermaidParser, plugins.pubGen], // pub-gen mermaid parser.

            [citePlugin, plugins.cite], // @see remark-cite.
            [citeParser, plugins.pubGen], // pub-gen citation parser, must be after citePlugin.

            [countParser, plugins.pubGen], // pub-gen count parser, must be the last pub-gen parser/compiler.

            [remarkStringify, plugins.stringfy],
            // [unifiedPrettier, plugins.prettier], // must be the last one.
        ].filter((plugin) => plugin !== null);

        const processor = unified();
        usedPlugins.forEach(([plugin, options]) => {
            processor.use(plugin, options);
        });
        return await processor.process(content);
    } catch (error) {
        const er = new Error(`Error processing content: ${error} ${content}`);
        console.error(error);
        return Promise.reject(error);
    }
};
