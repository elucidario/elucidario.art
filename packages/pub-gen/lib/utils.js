import path from "path";
import { getPaths, readFile } from "@elucidario/pkg-paths";
import { toMD } from "@elucidario/pkg-docusaurus-md";
import inquirer from "inquirer";
import { Console } from "@elucidario/pkg-console";
import { kebabCase } from "lodash-es";
import { fromMarkdown } from "mdast-util-from-markdown";

const pkg = readFile({
    filePath: path.resolve(getPaths().packages, "pub-gen", "package.json"),
}).value;

/**
 * Get publication config
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function pubGenConfig(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "pub-gen.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o pub-gen.json da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get publication package.json
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function packageJson(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "package.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o package.json da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get reference index of a publication
 *
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function referenceIndex(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "references",
                "index.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o index das referências da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get references of a publication
 *
 * @param {string} publication
 * @returns {Array}
 * @throws {Error}
 */
export function referencesFrom(publication) {
    try {
        return referenceIndex(publication).items.map((item) => {
            return readFile({
                filePath: path.resolve(
                    getPaths().references,
                    item.path.replace("<references>/", ""),
                ),
                ext: "json",
            }).value;
        });
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar as referências da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Generate a debounced function
 *
 * @param {function} func
 * @param {number} delay
 * @returns {function}
 */
export async function debounce(func, delay) {
    let timeoutId;
    return async (...args) => {
        clearTimeout(timeoutId);
        timeoutId = setTimeout(async () => {
            await func(...args);
        }, delay);
    };
}

/**
 * Clean falsy values from objects
 *
 * @param {object} obj
 * @returns {object}
 */
export function cleanFalsy(obj) {
    return Object.fromEntries(
        Object.entries(obj).reduce((acc, [key, value]) => {
            if (value !== null && value !== undefined && value !== "") {
                if (
                    typeof value === "object" &&
                    !Array.isArray(value) &&
                    Object.keys(value).length > 0
                ) {
                    value = cleanFalsy(value);
                    acc.push([key, value]);
                } else if (Array.isArray(value) && value.length) {
                    value = value.map((item) => {
                        if (typeof item === "object") {
                            return cleanFalsy(item);
                        }
                        return item;
                    });
                    acc.push([key, value]);
                } else if (typeof value === "string" && value.length) {
                    value = value.trim();
                    acc.push([key, value]);
                } else if (typeof value === "boolean" || value === 0) {
                    acc.push([key, value]);
                } else {
                    return acc;
                }
            }
            return acc;
        }, []),
    );
}

/**
 * Get publication package name
 *
 * @param {string} publication
 * @returns {string}
 */
export const publicationPkgName = (publication) =>
    `@elucidario/pub-${kebabCase(publication.trim())}`;

/**
 * Create a inquirer callback
 *
 * @param {string} title
 * @param {object} prompt
 * @param {function} callback
 * @returns {object}
 *
 * @example
 * const answers = async () => await makeInquirer({
 *    title: "Adding licenses...",
 *    prompt: pubGenPrompt("addLicense"),
 * });
 */
export async function makeInquirer({
    title,
    prompt,
    callback,
    publication,
    type = "info",
}) {
    let packageJSON = {};
    try {
        packageJSON = packageJson(publication);
    } catch (error) {
        packageJSON = pkg;
    }
    const console = new Console(packageJSON);
    console.log(title, { type });
    if (typeof callback === "undefined") {
        callback = async (answers) => {
            return cleanFalsy(answers);
        };
    }
    return await inquirer.prompt(prompt).then(callback);
}

/**
 * Creates the README.md file for the publication
 *
 * @param {string} name | name of the publication
 * @param {Object} answers | package.json of the publication
 * @param {string} answers.description | description of the publication
 *
 * @returns | README.md as string (needs to be saved)
 */
export function createREADME(name, answers) {
    return toMD([
        `# \`${publicationPkgName(name)}\`\n`,
        `> Publicação gerada com [pub-gen](https://elucidario.art/pub-gen)`,
        answers.description,
    ]);
}

/**
 * Creates the .gitignore file for the publication
 *
 * @returns | .gitignore as string (needs to be saved)
 */
export function createGitignore(files = []) {
    return [
        "node_modules",
        "credentials.json",
        "token.json",
        "dist",
        "~$*.*",
        ...files,
    ].join("\n");
}

/**
 * Creates the pub-gen.json file for the publication
 * use for current version schema
 *
 * @see package.json->config.schema-version for current version
 * @see packages/pub-gen/static/pub-gen/schemas/pub-gen-schema.json for schema
 *
 * @param {string} name | name of the publication
 * @param {Object} answers | answers from the prompt
 *
 * @returns | pub-gen.json as object (needs to be stringified to be saved)
 */
export function createPubGenJson(name, answers) {
    const { keywords, ...rest } = answers;
    const version = pkg.config["schema-version"];
    return cleanFalsy({
        $schema:
            "node_modules/@elucidario/pkg-pub-gen/static/pub-gen/schemas/pub-gen-schema.json",
        version,
        profile: "data-package",
        id: `https://elucidario.art/publicacoes/${name.trim()}`,
        ...rest,
        keywords:
            typeof keywords !== "undefined"
                ? keywords.split(/[;,]/).map((k) => k.trim())
                : undefined,
        created: new Date().toISOString(),
    });
}

/**
 * Creates the package.json file for the publication
 * @param {string} name | name of the publication
 * @returns | package.json
 */
export function createPackageJson(name) {
    const pubName = kebabCase(name.trim());
    return {
        name: publicationPkgName(name),
        version: "0.1.0",
        main: "README.md",
        private: true,
        scripts: {
            "add-author": `pub-gen add-author -p ${pubName}`,
            authenticate: `pub-gen authenticate -p ${pubName}`,
            build: `pnpm clean && pub-gen build -p ${pubName}`,
            clean: "rm -rf dist/*",
            convert: `pub-gen convert -p ${pubName}`,
            "ref-add": `pub-gen reference add -p ${pubName}`,
            "ref-search": `pub-gen reference search -p ${pubName}`,
            "version-up": `pub-gen version -p ${pubName}`,
        },
        devDependencies: {
            "@elucidario/pkg-pub-gen": "workspace:^",
        },
    };
}

/**
 * Parse node value
 *
 * @param {Object} node
 * @returns {Object}
 */
export function parseNodeValue(value) {
    const [action, optionsString] = value
        .replace("{{", "")
        .replace("}}", "")
        .split(":");

    const [filePath, ...fileOptions] = optionsString.split(";");

    const options = {};
    for (let option of fileOptions) {
        const [key, value] = option.split("=");
        options[key] = value;
    }

    if (action === "count") {
        return {
            action,
            optionsString,
            type: filePath,
            options,
        };
    }

    return {
        action,
        optionsString,
        filePath,
        fileOptions: options,
    };
}

/**
 *  Check if node value is a pub-gen node value ({{...}})
 * @param {string} value | node value
 * @returns | true if node value is a pub-gen node value ({{...}}), false otherwise
 */
export function isPubGenNodeValue(value) {
    return value.startsWith("{{") && value.endsWith("}}");
}

/**
 * Replace Regex Handlebars
 *
 * @param {string} text
 * @param {Object} options
 * @returns {string}
 */
export function replaceRegexHandlebars(text, options) {
    if (!text) return null;
    const regex = /{{(.*?)}}/g;
    const replacedText = text.replace(regex, (match, group) => {
        const replacement = options[group.trim()] || "";
        return replacement;
    });

    return replacedText;
}

/**
 *
 * @param {string} md | Markdown string
 * @param {object} options | options for parser
 * @param {boolean} options.reduce | reduce mdast tree
 * @returns
 */
export function mdToMdast(md, options = undefined) {
    if (typeof options === "undefined" || !options.reduce)
        return fromMarkdown(md);

    return fromMarkdown(md).children.reduce((acc, node) => {
        switch (node.type) {
            case "root":
            case "paragraph":
                acc = node.children;
                break;
        }
        return acc;
    }, []);
}

/**
 * Flatten style structure files
 * @param {object} data
 * @param {array} exclude
 * @param {array} paths
 * @returns {array}
 */
export function flattenStyleStructureFiles(data, exclude = [], paths = []) {
    for (const [key, value] of Object.entries(data)) {
        if (typeof value === "string" && value.endsWith(".md")) {
            paths.push(value);
        } else if (Array.isArray(value)) {
            for (const [name, include] of value) {
                if (
                    include &&
                    !exclude.includes(name) &&
                    !exclude.includes(key)
                ) {
                    paths.push(`${key}/${name}.md`);
                }
            }
        } else if (typeof value === "object") {
            flattenStyleStructureFiles(value, exclude, paths); // Recursively traverse
        }
    }
    return paths;
}

/**
 * Loop through style structure
 * @param {object} structure
 * @param {function} callback
 * @param {string} prefix
 * @returns {array}
 */
export async function loopStylesStructure(structure, callback, prefix = "") {
    for (const [key, value, title] of Object.entries(structure)) {
        if (typeof value === "string") {
            await callback({
                key: prefix ? prefix : key,
                value,
                title,
            });
        } else if (Array.isArray(value)) {
            for (const [name, required, title] of value) {
                await callback({
                    key: prefix ? prefix : key,
                    value: `${prefix ? `${prefix}/` : ""}${key}/${name}`,
                    required,
                    title,
                });
            }
        } else if (typeof value === "object") {
            await loopStylesStructure(value, callback, key); // Recursively traverse
        }
    }
}

/**
 * Process files
 * @param {object} obj
 * @param {function} callback
 * @returns {object}
 */
export function processFiles(obj, callback) {
    if (Array.isArray(obj)) {
        return obj.map((item) => processFiles(item, callback));
    } else if (typeof obj === "object") {
        if (
            obj.hasOwnProperty("file") &&
            obj.hasOwnProperty("path") &&
            obj.hasOwnProperty("processed")
        ) {
            return callback(obj);
        } else {
            return Object.fromEntries(
                Object.entries(obj).map(([key, value]) => {
                    return [key, processFiles(value, callback)];
                }),
            );
        }
    } else {
        return obj;
    }
}
