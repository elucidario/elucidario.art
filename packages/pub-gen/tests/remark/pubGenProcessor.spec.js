import path from "path";
import { pubGenProcessor } from "../../lib/remark/pubGenProcessor";
import {
    readContents,
    getPaths,
    readFile,
    dirExists,
    parseFile,
} from "@elucidario/pkg-paths";
import { packageJson } from "../../lib/utils";

describe("pubGenProcessor", () => {
    const content = readFile({
        filePath: path.resolve(
            getPaths().publications,
            "publicacao-teste",
            "content",
            "pt-br",
            "internal",
            "body",
            "hello-world.md",
        ),
        ext: "md",
    }).content;

    const mermaidNode = readFile({
        filePath: path.resolve(
            getPaths().publications,
            "publicacao-teste",
            "content",
            "pt-br",
            "internal",
            "body",
            "nodes",
            "mermaid-node.md",
        ),
        ext: "md",
    }).content;

    const pkg = packageJson("publicacao-teste");

    it("should rejects promise without content", async () => {
        expect.assertions(1);
        await expect(() => pubGenProcessor()).rejects.toThrow(
            "No content provided.",
        );
    });

    it("should rejects promise without options", async () => {
        expect.assertions(1);
        await expect(() => pubGenProcessor(content)).rejects.toThrow(
            "No options provided.",
        );
    });

    it("should rejects promise without options.publication", async () => {
        expect.assertions(1);
        await expect(() => pubGenProcessor(content, {})).rejects.toThrow(
            "No publication provided.",
        );
    });

    it("should rejects promise without options.lang", async () => {
        expect.assertions(1);
        await expect(() =>
            pubGenProcessor(content, {
                publication: "publicacao-teste",
            }),
        ).rejects.toThrow("No lang provided.");
    });

    it("should rejects promise without options.style", async () => {
        expect.assertions(1);
        await expect(() =>
            pubGenProcessor(content, {
                publication: "publicacao-teste",
                lang: "pt-br",
            }),
        ).rejects.toThrow("No style provided.");
    });

    it("should rejects promise without options.assets", async () => {
        expect.assertions(1);
        await expect(() =>
            pubGenProcessor(content, {
                publication: "publicacao-teste",
                lang: "pt-br",
                style: {
                    name: "abnt",
                    csl: "universidade-de-sao-paulo-escola-de-comunicacoes-e-artes-abnt.csl",
                },
            }),
        ).rejects.toThrow("No assets provided.");
    });

    it("should rejects promise without options.assetsTitles", async () => {
        expect.assertions(1);
        await expect(() =>
            pubGenProcessor(content, {
                publication: "publicacao-teste",
                lang: "pt-br",
                style: {
                    name: "abnt",
                    csl: "universidade-de-sao-paulo-escola-de-comunicacoes-e-artes-abnt.csl",
                },
                assets: {},
            }),
        ).rejects.toThrow("No assetsTitles provided.");
    });

    it("should rejects promise without options.pkg", async () => {
        expect.assertions(1);
        await expect(() =>
            pubGenProcessor(content, {
                publication: "publicacao-teste",
                lang: "pt-br",
                style: {
                    name: "abnt",
                    csl: "universidade-de-sao-paulo-escola-de-comunicacoes-e-artes-abnt.csl",
                },
                assets: {},
                assetsTitles: {
                    mermaid: "Diagrama",
                },
            }),
        ).rejects.toThrow("No package.json provided.");
    });

    it("should return processed value", async () => {
        const processed = await pubGenProcessor(content, {
            publication: "publicacao-teste",
            lang: "pt-br",
            style: {
                name: "abnt",
                csl: "universidade-de-sao-paulo-escola-de-comunicacoes-e-artes-abnt.csl",
            },
            assets: {},
            assetsTitles: {
                mermaid: "Diagrama",
            },
            pkg,
        });

        expect(processed.value).toMatchSnapshot();
    });
});
