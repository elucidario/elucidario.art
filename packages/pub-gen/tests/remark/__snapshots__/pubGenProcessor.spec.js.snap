// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pubGenProcessor should return processed value 1`] = `
"# Hello World

## Second level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var (Linked Art, 2021a).

**undefined 1: Mermaid**

\`\`\`{.mermaid filename=mermaid loc="C:\\\\Users\\\\55119\\\\Elucidário.art\\\\elucidario\\\\publications\\\\publicacao-teste\\\\files\\\\generated" width=720 background=transparent}
erDiagram
        OBJETO ||--|{ CLASSIFICACAO : ""
\`\`\`

**Fonte:** Elaborado pelo autor, com base em [@caro-castro2011]

### Third level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. (Linked Art, 2021a)

#### Fourth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. (Linked Art, 2021b)

##### Fifth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. (Linked Art, 2021a, 2021b)

###### Sixth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var.

**undefined 2: Lorem ipsum**

![Lorem ipsum](C:\\Users\\55119\\Elucidário.art\\elucidario\\publications\\publicacao-teste\\files\\static\\lorem-ipsum.jpeg)
"
`;
